### PR TITLE
Polars lazyframe

### PR DESCRIPTION
**Pull request by cosmicBboy**
## Tracking issue

Fixes https://github.com/flyteorg/flyte/issues/5678

## Why are the changes needed?

The flytekit plugin for polars only supports `polars.DataFrame`. Many polars users rely on the `LazyFrame` API to make data handling more memory efficient.

## What changes were proposed in this pull request?

This PR:
- Adds and registers a `LazyFrame` encoder/decoder to the `StructuredDataset` type.

## How was this patch tested?

Unit tests run on CI.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 1]
```diff
@@ -14,20 +14,33 @@
...
     """
 
-    def to_html(self, df: pl.DataFrame) -> str:
-        assert isinstance(df, pl.DataFrame)
-        describe_df = df.describe()
-        return pd.DataFrame(describe_df.transpose(), columns=describe_df.columns).to_html(index=False)
+    def to_html(self, df: typing.Union[pl.DataFrame, pl.LazyFrame]) -> str:
+        assert isinstance(df, (pl.DataFrame, pl.LazyFrame))
+        if isinstance(df, pl.LazyFrame):
+            describe_df = df.collect().describe()
```
Comment (thomasjpfan) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:37]: `describe` does not need to be collected first with a lazy dataframe:

```python
import polars as pl

df = pl.DataFrame({"A": [1, 2, 3], "b": [4, 5, 6]}).lazy()
assert isinstance(df, pl.LazyFrame)

df.describe()
```

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 2]
```diff
@@ -82,6 +95,76 @@ def decode(
...
+        # Polars 0.13.12 deprecated to_parquet in favor of write_parquet
+        if hasattr(df, "write_parquet"):
+            df.write_parquet(output_bytes)
+        else:
+            df.to_parquet(output_bytes)
+
+        if structured_dataset.uri is not None:
+            fs = ctx.file_access.get_filesystem_for_path(path=structured_dataset.uri)
+            with fs.open(structured_dataset.uri, "wb") as s:
+                s.write(output_bytes)
```
Comment (thomasjpfan) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:127]: Can we write directly into the cloud storage and not use `io.BytesIO()`?

```python
with fs.open(structured_dataset.uri, "wb") as s:
    df.write_parquet(s)
```


REF: https://docs.pola.rs/user-guide/io/cloud-storage/

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py (hunk 1 above)
Comment (cosmicBboy) [reply plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:37]: it does for polars <= 0.19. Unless we place a lower bound on the polars version supported by this plugin.

Comment (thomasjpfan) [reply plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:37]: Can we try `describe` first and fallback to `collect` if it does not work?

```python
try:
    describe_df = df.describe()
except AttributeError:
    # LazyFrames in polars <= 0.19 does not support `describe` 
    describe_df = df.collect().describe()
```

**Review by eapolinario — DISMISSED** [review]

**Review by thomasjpfan — COMMENTED** [review]
On the flyte sandbox, the following stalls for me:

```python
from flytekit import task, workflow, ImageSpec
import polars as pl

polars_plugin = "git+https://github.com/flyteorg/flytekit@55db243b51bfe5137942b68aa212ecb8edc27bcd#subdirectory=plugins/flytekit-polars"

image = ImageSpec(
    name="polars",
    packages=[polars_plugin, "flytekit"],
    registry="localhost:30000",
    apt_packages=["git"],
)


@task(container_image=image)
def create_lazy_frame() -> pl.LazyFrame:
    return pl.LazyFrame({"col1": [1, 3, 2], "col2": list("abc")})


@task(container_image=image)
def read_pl(df: pl.LazyFrame) -> int:
    return df.select(pl.col("col1").sum()).collect().item()


@workflow
def main() -> int:
    df = create_lazy_frame()
    return read_pl(df=df)

```

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 3]
```diff
@@ -82,6 +98,79 @@ def decode(
...
+
+# Register the Polars DataFrame handlers
 StructuredDatasetTransformerEngine.register(PolarsDataFrameToParquetEncodingHandler())
 StructuredDatasetTransformerEngine.register(ParquetToPolarsDataFrameDecodingHandler())
 StructuredDatasetTransformerEngine.register_renderer(pl.DataFrame, PolarsDataFrameRenderer())
+
+# Register the Polars LazyFrame handlers
+StructuredDatasetTransformerEngine.register(PolarsLazyFrameToParquetEncodingHandler())
+StructuredDatasetTransformerEngine.register(ParquetToPolarsLazyFrameDecodingHandler())
+StructuredDatasetTransformerEngine.register_renderer(pl.LazyFrame, PolarsDataFrameRenderer())
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:176]: should we make the renderer for this explicit? (either using an `Annotated` type or in the task function body?)

**Comment by cosmicBboy** [comment]
[failing greatexpectations plugin](https://github.com/flyteorg/flytekit/actions/runs/10530626765/job/29181131760?pr=2695) should be fixed in a separate PR

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py]: @thomasjpfan I removed `register_renderer` for the LazyFrame... I think it's reasonable to think that any user who uses this doesn't want to perform potentially computationally intensive operation at the end of the task without doing it explicitly

**Comment by eapolinario** [comment]
@cosmicBboy , @thomasjpfan , can we merge this?

**Comment by cosmicBboy** [comment]
one sec @eapolinario let's not merge yet, still testing on flyte sandbox and serverless

**Review by pingsutw — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]

**Comment by HansBambel** [comment]
Cool feature!
Is the documentation then just not up to date? There is no mention of polars at all in the type-mapping:

https://docs.flyte.org/en/latest/user_guide/data_types_and_io/index.html#python-to-flyte-type-mapping

**Comment by kumare3** [comment]
Yes @HansBambel - would you mind adding an update

**Comment by HansBambel** [comment]
@kumare3 Yes, opened a PR for this.

**Comment by HansBambel** [comment]
For better understanding: Even though `LazyFrame` is supported, using it across tasks does not yield any benefits such as low memory usage and lazy evaluation as `StructuredDataset` calls `.collect()` to serialize it.

This results in using `LazyFrame` over `DataFrame` when dealing with huge files that do not fit in memory useless, right?

**Comment by cosmicBboy** [comment]
@HansBambel correct. IIRC streaming a LazyFrame to blob store was not supported at the time of this PR (I believe this is still the case). Tried using `sink_parquet`: https://docs.pola.rs/api/python/stable/reference/api/polars.LazyFrame.sink_parquet.html but it's not considered stable and didn't work with fsspec (neither did `scan_parquet`).

**Pull request merged**